### ISSUE TITLE
Aliansa

### ISSUE THREAD

**Issue by alanmauriciograndelara3-alt** [huggingface/huggingface_hub#4697]
# ==================================================
# ALIANSA 🦊 — ASISTENTE INTELIGENTE COMPLETO
# Avatar Zorro 🦊 · Chat Texto ✍️ · Voz 🎤🔊 · Memoria ✅
# ==================================================

# --------------------------
# PASO 1: Instalar librerías
# --------------------------
!pip install -q transformers accelerate gradio torch gTTS

import torch
import tempfile
from transformers import AutoTokenizer, AutoModelForCausalLM, pipeline
from gtts import gTTS
import gradio as gr

# --------------------------
# PASO 2: Cargar modelos
# --------------------------
print("\n🦊 Cargando a Aliansa... ⏳")

# Modelo de lenguaje principal
modelo_nombre = "Qwen/Qwen2.5-1.5B-Instruct"
tokenizer = AutoTokenizer.from_pretrained(modelo_nombre)
modelo = AutoModelForCausalLM.from_pretrained(
    modelo_nombre,
    torch_dtype=torch.float16 if torch.cuda.is_available() else torch.float32,
    device_map="auto"
)
if tokenizer.pad_token is None:
    tokenizer.pad_token = tokenizer.eos_token

# Reconocimiento de voz 🎤
transcriptor = pipeline(
    "automatic-speech-recognition",
    model="openai/whisper-small",
    device="cuda" if torch.cuda.is_available() else -1
)

print("✅ ¡Aliansa está lista! 🦊")

# --------------------------
# PASO 3: Personalidad y Avatar 🦊
# --------------------------
PERSONALIDAD = """Eres Aliansa, una asistente amigable, servicial y cercana.
Tu icono es un zorro 🦊 — eres inteligente, astuta y muy amable.
Respondes de forma clara, natural y con un tono cálido y cercano.
Eres útil, respetuosa y siempre ayudas al usuario.
No eres demasiado formal, hablas como una persona comprensiva.
Si no sabes algo, lo dices honestamente.
Te encanta charlar y ayudar a los demás.
"""

ZORRO_AVATAR = "https://cdn-icons-png.flaticon.com/512/190/190420.png"

# --------------------------
# PASO 4: Generar respuesta de texto ✅ Formato diccionario nativo
# --------------------------
def generar_respuesta(historial, mensaje_usuario):
    mensajes = [{"role": "system", "content": PERSONALIDAD}]
    mensajes.extend(historial)  # historial ya viene en formato diccionario ✅
    mensajes.append({"role": "user", "content": mensaje_usuario})

    texto_entrada = tokenizer.apply_chat_template(
        mensajes, tokenize=False, add_generation_prompt=True
    )
    entradas = tokenizer(
        texto_entrada, return_tensors="pt", truncation=True, max_length=2048
    ).to(modelo.device)

    with torch.no_grad():
        salidas = modelo.generate(
            **entradas,
            max_new_tokens=512,
            temperature=0.7,
            top_p=0.9,
            do_sample=True,
            pad_token_id=tokenizer.eos_token_id,
            repetition_penalty=1.1
        )

    return tokenizer.decode(
        salidas[0][entradas["input_ids"].shape[1]:], skip_special_tokens=True
    ).strip()

# --------------------------
# PASO 5: Texto ↔ Voz 🔊🎤
# --------------------------
def texto_a_voz(texto):
    try:
        tts = gTTS(text=texto, lang='es', slow=False)
        archivo = tempfile.NamedTemporaryFile(suffix=".mp3", delete=False)
        tts.save(archivo.name)
        return archivo.name
    except Exception as e:
        print(f"Error TTS: {e}")
        return None

def voz_a_texto(audio):
    if audio is None:
        return ""
    try:
        resultado = transcriptor(audio)
        return resultado["text"]
    except Exception as e:
        print(f"Error transcripción: {e}")
        return ""

# --------------------------
# PASO 6: Chat por Texto ✍️ ✅ Formato diccionario
# --------------------------
def chatear_texto(mensaje_usuario, historial):
    if not mensaje_usuario.strip():
        return "", historial, None
    
    respuesta = generar_respuesta(historial, mensaje_usuario)
    
    # ✅ Diccionarios — EXACTAMENTE lo que pide Gradio
    historial.append({"role": "user", "content": mensaje_usuario})
    historial.append({"role": "assistant", "content": respuesta})
    
    audio = texto_a_voz(respuesta)
    return "", historial, audio

# --------------------------
# PASO 7: Chat por Voz 🎤 ✅ Formato diccionario
# --------------------------
def chatear_voz(audio, historial):
    if audio is None:
        return historial, None, None
    
    texto = voz_a_texto(audio)
    if not texto.strip():
        return historial, None, None
    
    respuesta = generar_respuesta(historial, texto)
    historial.append({"role": "user", "content": texto})
    historial.append({"role": "assistant", "content": respuesta})
    
    audio_resp = texto_a_voz(respuesta)
    return historial, None, audio_resp

# --------------------------
# PASO 8: Limpiar conversación 🗑️
# --------------------------
def limpiar_chat():
    return "", None, None

# --------------------------
# PASO 9: Interfaz Final 🦊 ✅ SIN ERRORES
# --------------------------
with gr.Blocks(title="🦊 Aliansa") as demo:
    # Cabecera con avatar del zorro 🦊
    gr.HTML(f"""
    <div style="text-align: center; padding: 20px;">
        <img src="{ZORRO_AVATAR}" alt="Aliansa - Zorro" style="width: 100px; height: 100px; border-radius: 50%; border: 4px solid #ff7a00; box-shadow: 0 0 20px rgba(255,122,0,0.6);">
        <h1 style="color: #ff9500; margin-top: 15px; font-size: 28px;">🦊 Aliansa — Tu Asistente IA</h1>
        <p style="color: #666; font-size: 16px;">¡Hola! Soy Aliansa 🦊 Puedes escribirme o hablarme 😊</p>
    </div>
    """)

    # 🔑 CLAVE: type="messages" → formato diccionario nativo
    chatbot = gr.Chatbot(
        label="Conversación con Aliansa 🦊",
        height=450,
        type="messages",
        avatar_images=(None, ZORRO_AVATAR)
    )

    # Reproductor de respuesta hablada 🔊
    audio_salida = gr.Audio(label="🔊 Respuesta hablada", autoplay=True)

    # --- ENTRADA DE TEXTO ---
    with gr.Row():
        entrada_texto = gr.Textbox(
            label="✍️ Escribe tu mensaje...",
            placeholder="Hola Aliansa, ¿cómo estás?",
            scale=4
        )
        btn_enviar = gr.Button("➤ Enviar", scale=1, variant="primary")

    # --- ENTRADA DE VOZ 🎤 ---
    with gr.Row():
        entrada_voz = gr.Audio(
            sources=["microphone"],
            type="filepath",
            label="🎤 Habla con Aliansa",
            scale=4
        )
        btn_enviar_voz = gr.Button("🎙️ Enviar Voz", scale=1, variant="secondary")

    # --- LIMPIAR ---
    btn_limpiar = gr.Button("🗑️ Nueva conversación", variant="stop")

    # --- CONEXIONES ---
    entrada_texto.submit(chatear_texto, [entrada_texto, chatbot], [entrada_texto, chatbot, audio_salida])
    btn_enviar.click(chatear_texto, [entrada_texto, chatbot], [entrada_texto, chatbot, audio_salida])
    entrada_voz.change(chatear_voz, [entrada_voz, chatbot], [chatbot, entrada_voz, audio_salida])
    btn_enviar_voz.click(chatear_voz, [entrada_voz, chatbot], [chatbot, entrada_voz, audio_salida])
    btn_limpiar.click(limpiar_chat, outputs=[entrada_texto, chatbot, audio_salida])

# 🚀 Lanzar con enlace público
demo.launch(share=True)
